### ISSUE TITLE
Program linked to libgrpc++.so.1(doesn't exists) instead if libgrpc++.so.4. 

### ISSUE THREAD

**Issue by david9991** [grpc/grpc#11868]
### Should this be an issue in the gRPC issue tracker?
Yes
 
### What version of gRPC and what language are you using?
 v1.4.2
 
### What operating system (Linux, Windows, …) and version?
 Ubuntu 16.04
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 gcc, g++; C++
 
### What did you do?
 1. Install grpc via linuxbrew.
 2. Compile my program with grpc library.
 
### What did you expect to see?
  Program can run with libgrpc++.so.4
 
### What did you see instead?
  When I run the program, it said that couldn't find the library `libgrpc++.so.1`, but I just have `libgrpc++.so.4`. And I tried to search text `libgrpc++.so.1` in file `libgrpc++.so.4`. yes, it have a match (but it shouldn't).

**Comment by sgleizes** [comment]
Having the same issue here. The last release without it that I have tried is 1.1.4.

**Comment by david9991** [comment]
May be the reason is
```
$(Q) $(LDXX) $(LDFLAGS) -L$(LIBDIR)/$(CONFIG) -shared -Wl,-soname,libgrpc++.so.1 -o $(LIBDIR)/$(CONFIG)/libgrpc++$(SHARED_VERSION_CPP).$(SHARED_EXT_CPP) $(LIBGRPC++_OBJS) $(ZLIB_MERGE_LIBS) $(CARES_MERGE_LIBS) $(LDLIBSXX) $(LDLIBS_PROTOBUF) $(LDLIBS) -lgrpc -lgpr
$(Q) ln -sf $(SHARED_PREFIX)grpc++$(SHARED_VERSION_CPP).$(SHARED_EXT_CPP) $(LIBDIR)/$(CONFIG)/libgrpc++$(SHARED_VERSION_CPP).so.1
$(Q) ln -sf $(SHARED_PREFIX)grpc++$(SHARED_VERSION_CPP).$(SHARED_EXT_CPP) $(LIBDIR)/$(CONFIG)/libgrpc++$(SHARED_VERSION_CPP).so
```
Used for compile but 
```
$(Q) ln -sf $(SHARED_PREFIX)grpc++$(SHARED_VERSION_CPP).$(SHARED_EXT_CPP) $(prefix)/lib/libgrpc++.so.4
$(Q) ln -sf $(SHARED_PREFIX)grpc++$(SHARED_VERSION_CPP).$(SHARED_EXT_CPP) $(prefix)/lib/libgrpc++.so
```
for installation.

`-Wl,-soname,libgrpc++.so.1` should be `-Wl,-soname,libgrpc++.so.4` and `libgrpc++.so` should link to `libgrpc++.so.4` isn't it.

**Comment by BryceCicada** [comment]
Same issue here still on 1.4.5.

Strange that this is tagged as an enhancement.  Any application linking to grpc++ will crash at runtime.  Is this not a bug?

Right now, I've just worked around by soft-linking like ```ln -s libgrpc++.so.1.4.5 libgrpc++.so.1```

**Comment by BryceCicada** [comment]
There's something strange here.  It works on Ubuntu 16.04, but not on Centos 7.  

Here's some Dockerfiles....

The following works ok:
```
FROM ubuntu:16.04

# Update environment and install dependencies
RUN apt-get update && \
    apt-get -y install \
  	build-essential \
	autoconf \
	libtool \
	libgflags-dev \
	libgtest-dev \
	clang \
	libc++-dev \
	curl \
	git && \
    apt-get clean

# Clone latest GRPC, build and install  
RUN cd /usr/local/src && \
    git clone -b v1.4.2 https://github.com/grpc/grpc && \
    cd grpc && \
    git submodule update --init && \
    make -j 8 && \
    make install

# Create a simple test application that links against grpc++ and crashes at runtime
RUN cd /usr/local/src && \
    mkdir test-app && \
    echo "int main() {}" > main.cpp && \
    g++ -L/usr/local/lib -lgrpc++ main.cpp -o main && \
    ./main
```

The following fails:
```
FROM centos:7

# Update environment and install dependencies
RUN yum update -y && \
    yum install -y \
  	build-essential \
	autoconf \
	libtool \
	gflags-develop \
	gtest-develop \
	clang \
	libstdc++-devel \
	curl \
	git \
	make \ 
	gcc \
	gcc-c++ \
	which && \
    yum clean all
  
# Clone latest GRPC, build and install  
RUN cd /usr/local/src && \
    git clone -b v1.4.2 https://github.com/grpc/grpc && \
    cd grpc && \
    git submodule update --init && \
    make -j 8 && \
    make install

# Create a simple test application that links against grpc++ and crashes at runtime
RUN cd /usr/local/src && \
    mkdir test-app && \
    echo "int main() {}" > main.cpp && \
    g++ -L/usr/local/lib -lgrpc++ main.cpp -o main && \
    ./main

```

**Comment by david9991** [comment]
Upgraded to latest version, the issue seems gone.

**Comment by sgleizes** [comment]
It's because it has never been fixed: #13500 is still pending.

**Comment by jtattermusch** [comment]
This was fixed by https://github.com/grpc/grpc/pull/16832. The fix will be in the upcoming v1.17.0 release.

Also basically a duplicate of https://github.com/grpc/grpc/issues/10299
